### ISSUE TITLE
automatic differentiation?

### ISSUE THREAD

**Issue by alejandroschuler** [vincentarelbundock/marginaleffects#749]
Hey, I'm new here but have a quick question regarding the computation of required derivatives. I understand this is done with finite differences (more or less). 

Is there a reason the package shouldn't exploit modern autodifferentiation libraries? I'm not sure what the state-of-the-art is for that in R but the required libraries are super well-developed and used everywhere in julia and python and surely there are linkages. 

Anyways, thanks for the great package- I use it in teaching!

**Comment by vincentarelbundock** [comment]
Glad you like the package!

We already support the Richardson algorithm via the ˋnumDeriv` package. Adding support for other methods would be almost trivial, since the infrastructure is all there. All one would have to do is modify this file:

https://github.com/vincentarelbundock/marginaleffects/blob/main/R/get_jacobian.R

Note that x is a vector of coefficients and func(x) returns a vector of estimates/predictions/comparisons/slopes.

I'd be super happy to merge a pull request. If you have the time but have a specific autodiff package to recommend in R, I can also look into it.

The main advantage of finite differences is obviously speed. Each call to func() can take a few seconds in large datasets/models. And fancy autodiff strategy often require a lot of evaluations of the function, which is more reasonable to do when the function is cheaper.

**Comment by alejandroschuler** [comment]
Sure, here are a few that I know of:
- https://tensorflow.rstudio.com/guides/tensorflow/autodiff
- https://github.com/shabbychef/madness
- https://non-contradiction.github.io/autodiffr/articles/autodiffr_intro.html

To be clear, what I mean by "[autodiff](https://en.wikipedia.org/wiki/Automatic_differentiation)" is not numeric differentiation. So there are actually no calls to the function at all- such a system symbolically computes the derivative as a function and then to find the derivative it evaluates that derivative function once. There is therefore no error (up to machine precision) in the derivative and if the derivatives are precompiled (as they could be for the standard parameters) computation would be nothing but a single function evaluation.

**Comment by vincentarelbundock** [comment]
Ah yes, I misread. That's what I get for commenting on Github issues at 6AM.

I don't have experience with any of those packages. Will look into them.

**Comment by alejandroschuler** [comment]
Also fyi i just asked chatGPT to write a native R autodiff library prototype and it looked pretty good so that's another option! :)

**Comment by etiennebacher** [comment]
I don't know much about this but I saw Dirk's tweet on this RcppFastAD package a few months ago: https://github.com/eddelbuettel/rcppfastad

I hope it's useful

**Comment by vincentarelbundock** [comment]
I have not been able to spend much time on this, but my sense is that this is likely to be very difficult. For example, the `autodiffr` package [does not support even some of the most basic R operators](https://non-contradiction.github.io/autodiffr/articles/autodiffr_intro.html#basic-r-functions-that-have-problems-with-autodiffr), so it’s likely that we would have to hand-code a bunch of prediction functions. I’m not sure how to do that in a way that would be robust to all possible models, and scalable, especially models with interactions or in-formula transformations.

Below, I leave a super simple example with both automatic and manual computation of the average slope by finite difference.

Perhaps someone will be able to use this as the basis for a proof of concept with one of the packages above. In addition to the slope shown below, we would need the package to generate a jacobian matrix of the slopes with respect to the parameters, and then we could use that to compute the standard errors of the slopes.

``` r
library(marginaleffects)
mod = lm(mpg ~ wt * log(qsec), data = mtcars)

# automatic
avg_slopes(mod, variables = "qsec")
# 
#  Term Estimate Std. Error     z   Pr(>|z|)  2.5 % 97.5 %
#  qsec   0.8992     0.2727 3.298 0.00097424 0.3648  1.434
# 
# Prediction type:  response 
# Columns: type, term, estimate, std.error, statistic, p.value, conf.low, conf.high

# manual
dat_hi = transform(mtcars, qsec = qsec + 1e-4 / 2)
dat_lo = transform(mtcars, qsec = qsec - 1e-4 / 2)
X_hi = model.matrix(mod, data = dat_hi)
X_lo = model.matrix(mod, data = dat_lo)
pred_hi = drop(X_hi %*% coef(mod))
pred_lo = drop(X_lo %*% coef(mod))
mean(pred_hi - pred_lo) / 1e-4
# [1] 0.8991987
``

**Comment by vincentarelbundock** [comment]
Here's a toy problem to illustrate the kind of function we need to the get the jacobian for using autodiff.

```r
library(marginaleffects)

dat <- transform(iris, bin = as.numeric(Petal.Width > median(Petal.Width)))
dat <- get_dataset("military")
fml <- officer ~ rank * (gender + branch)
mod <- glm(fml, data = dat, family = binomial)

# Define a function f: R^p -> R^n
f <- function(beta) {
    X <- model.matrix(fml, data = dat)
    out <- as.vector(plogis(X %*% beta))
    return(out)
}

# Get Jacobian by forward diff
forward_diff <- function(f, beta0, eps = 0.0001) {
    J <- matrix(NA_real_, ncol = length(beta0), nrow = nrow(dat))
    for (i in seq_along(beta0)) {
        beta_tmp <- beta0
        beta_tmp[i] <- beta_tmp[i] + eps
        J[, i] <- (f(beta_tmp) - f(beta0)) / eps
    }
    return(J)
}

J <- forward_diff(f, coef(mod))
J |> head()

# Compare to built-in marginaleffects
predictions(mod, type = "response") |>
    attr("jacobian") |>
    unname() |>
    head()
```

**Comment by vincentarelbundock** [comment]
Well, that kind of speed up would be amazing, of course.

Now, obviously, `predictions()` is doing **a lot** of other sanity checks and processing here. And your computation is super fast, so before we get all excited, maybe we should try a model with many parameters. Like with `factor(x)` where `x` is random draws form `letters` so we have >=26 coefficients.

Interestingly, you might be able to plug the mechanism you have just demonstrated directly into `marginaleffects` to get full generality immediately. All you'd need to do is write a `get_jacobian_jax()` function and dispatch it within `get_jacobian()` here:

https://github.com/vincentarelbundock/marginaleffects/blob/f914d6aafeb4f8cf97e6b92f83200f5fe2b3793f/R/get_jacobian.R#L1

That feels like a very easy way to make an apples-to-apples comparison, including all the overhead that we actually want/need (user-input checks, data extraction and processing, etc.).

And if it works and provides meaningful speed gains there, then we have an immediate "win" for the entire `marginaleffects` ecosystem. If that works, it's a big deal!

**Comment by arcruz0** [comment]
### TL;DR

Benchmarks indicate massive gains vs numerical differentiation, but at least for raw `predictions()`, the overall gains would be relatively small (see the next section for details).

Unfortunately, generality would not be straightforward. The catch with autodiff is that you need to use specific functions that work with it—JAX provides many, mainly via `jax.numpy` and `jax.scipy`. So I think that a full JAX backend for `marginaleffects` would probably require many custom "output functions" (such as prediction functions; not sure about others) that use JAX-compatible parts.

I still find the premise appealing overall. For example, in the final section I show very quick SEs for average predictions in a linear model with N=10,000, K=126. This appeals to me as a possible way to, e.g., get analytic uncertainty estimates for predictions in `fixest` (#1487). 

@vincentarelbundock, if I wanted to pursue this further, what would you recommend? Would it make sense to try and develop a backend for `marginaleffects`? (Perhaps starting with a few popular models). Or should I just keep it as a separate function/package for occasions in which a super-lean approach is needed?

### Better benchmarks

> maybe we should try a model with many parameters. Like with `factor(x)` where `x` is random draws form `letters` so we have >=26 coefficients.

The previous model had 100 coefficients. The following benchmarks use one with 1000 coefficients (i.e., same code as before, but now with `K <- 1000`).

> Now, obviously, `predictions()` is doing **a lot** of other sanity checks and processing here. 

Certainly. Perhaps the following is a better benchmark. Let's say that from a linear model we have `X` and `beta`. I think that what `marginaleffects::predictions()` is now doing with the Jacobian boils down to the following (reconstructed from `get_jacobian()` and `get_predict.lm()`, omitting all checks):

```r
f_jacobian_raw <- function(X, beta){
  marginaleffects:::get_jacobian_fdforward(func = \(x) marginaleffects:::eigenMatMult(X, x), x = coef)
}
```

Let's see benchmarks for `predictions()`, the raw `get_jacobian_fdforward()` approach, and the JAX approach:

```r
microbenchmark::microbenchmark(
  me = predictions(m),
  numDeriv = f_jacobian_raw(mm, b),
  jax = f_jax_predictions_jacobian(mm, b),
  times = c(1, 1, 10)
)
#> Unit: milliseconds
#>     expr        min         lq      mean     median         uq        max neval
#>       me 24329.9987 24329.9987 24329.999 24329.9987 24329.9987 24329.9987     1
#> numDeriv  5597.5517  5597.5517  5597.552  5597.5517  5597.5517  5597.5517     1
#>      jax   221.5748   222.4675   229.696   225.7759   232.9505   250.1038    10
```

If I'm reading this correctly, autodiff is clearly faster than numerical differentiation, but making this change could cut only about a fifth of the execution time of `predictions()`. The bulk of time seems to be spent elsewhere (perhaps in how `newdata` is called? I can't quite pin it down from profiling).

I should also mention that my original motivation for trying this approach was not speed, but memory usage. Here JAX also appears to be a bit better than the numerical approach (even with the `reticulate` conversions in between). But that's again not a large part of the overall RAM usage of `predictions()`.

```r
peakRAM::peakRAM(
  predictions(m),
  f_jacobian_raw(mm, b),
  f_jax_predictions_jacobian(mm, b)
)
#>                      Function_Call Elapsed_Time_sec Total_RAM_Used_MiB Peak_RAM_Used_MiB
#> 1                   predictions(m)           22.112              238.6            1434.6
#> 2             f_jacobian_raw(mm,b)            5.509               76.3             236.8
#> 3 f_jax_predictions_jacobian(mm,b)            0.151               76.3              76.3
```

### Another example: average predictions

Here I added `d$letters <- sample(letters, N, replace = T)` before running the model, which increases the number of coefficients to 1026. Running `predictions(m, by = "letters")` takes ~70s. I don't know exactly how to recreate a "lean" version (without the checks, etc.). But here a JAX approach takes ~0.3s to get the SEs for the average predictions:

```r
microbenchmark::microbenchmark(jax_avg_pred = {
  f_jax_by <- function(X, coef, by){
    f_py <- \(x){
      # get predictions
      preds <- jax$numpy$stack(mm %*% x)
      # compute means by group
      g <- jax$numpy$unique(by, return_inverse = T)[[2]]
      jax$numpy$hstack(jax$ops$segment_sum(preds, g)) / jax$numpy$bincount(g)
    }
    # create Jacobian f and run it
    f_jacobian <- jax$jacfwd(f_py)
    J_jax <- f_jacobian(coef)
    # convert result
    np$array(J_jax)
  }
  
  # execute function
  groups <- as.factor(d$letters) |> as.integer() |> as.numeric()
  J <- f_jax_by(X, coef = np_array(coef(m)), by = np_array(groups))
  # compute prediction standard errors
  V <- vcov(m)
  ses <- diag(J %*% V %*% t(J)) |> sqrt()
  ses
}, times = 10)
#> Unit: milliseconds
#>          expr      min       lq     mean   median       uq      max neval
#>  jax_avg_pred 257.8643 303.9673 310.9641 311.6418 332.9921 341.9627    10
```

**Comment by vincentarelbundock** [comment]
Wow, this is very enticing. 

If we want to look at this seriously, I wonder if it wouldn't make sense to start from the Python version of `marginaleffects`.

1. The code base is much simpler because it supports fewer models and edge cases
2. It supports `PyFixest` which is similar to the R version: https://py-econometrics.github.io/pyfixest/pyfixest.html

There, I can imagine defining an alternative `inner()` function that would be triggered for a subset of calls where there are JAX functions already implemented.

https://github.com/vincentarelbundock/pymarginaleffects/blob/main/marginaleffects/predictions.py#L135

Then, it would be a matter of gradually implementing more JAX functions to cover more of the top-level functionality.
